src/ast.rs

```diff
@@ -195,12 +196,18 @@ pub enum Pat {
     Constant(BigInt),
 }
 
-#[derive(Debug, Clone, Encode, Decode)]
+#[derive(Debug, Clone, Encode, Decode, Eq, Hash)]
 pub struct TPat {
     pub v: Pat,
     pub t: Option<Type>,
 }
 
+impl PartialEq for TPat {
+    fn eq(&self, other: &Self) -> bool {
+        self.v == other.v
```

Comment:
I'm thinking that if https://github.com/anoma/vamp-ir/pull/27 is merged into `main`, then the `t: Type` field should only contain `None`. Hence we could just `#[derive(Eq, PartialEq)]`. My minor concern is that there are expressions that can have several different types assigned to them, and making them compare equal could be debatable.

Reply:
I am thinking this over and perhaps deriving PartialEq and Eq for TExpr etc is using a sledgehammer to drive a tack, and I can code around this pretty easily. I only use it to remove identities like `c1 = c2` or `v1 = v2` and I can test those equalities easily enough. My earlier code needed to check equality between entire expressions but now I avoid needing to do those choices in the first place.

Reply:
Also deriving `Hash` is no longer necessary. I think `bincode::Encode` or `bincode::Decode` must give `Hash` for free.